[#30] 주어진 위경도 범위내의 가게를 조회하는 기능 추가

canlion opened the pull request:
-

canlion committed:
feat: 가게 엔티티에 위경도 정보를 추가

* 가게 엔티티에 위경도 정보를 추가
  - 가게 등록: 주소를 네이버 geocoding 활용해 위경도로 변경, 엔티티에
    추가

canlion committed:
feat: 특정 위경도 범위 내의 가게 조회 기능 추가

* 주어진 위경도 범위 내에서 등록된 가게를 조회하는 기능 추가

canlion committed:
feat: 가게 조회 조건에 음식 필터를 추가

* 가게 검색 조건 음식 필터를 추가, 사용자가 원하는 형태의 음식을
  판매하는 가게를 검색할 수 있도록 함.

f-lab-lyan commented on api/src/main/java/burgermap/controller/StoreController.java:
```diff
@@ -52,17 +61,33 @@ public ResponseEntity<StoreInfoDto> getStore(@PathVariable Long storeId) {
-    public ResponseEntity<List<StoreInfoDto>> getStores() {
-        return ResponseEntity.ok(null);
+    @PostMapping("/search")
+    public ResponseEntity<StoreSearchResultDto> getStores(
```
컨트롤러에 비지니스로직이 들어가게 된 것 같아요.

f-lab-lyan commented on api/src/main/java/burgermap/service/GeoCodingService.java:
```diff
@@ -0,0 +1,83 @@
+                                .build())
+                .retrieve()
+                .bodyToMono(String.class)  // JSON 형식의 응답을 String으로 변환
+                .block();
```
비동기 API를 동기로 쓰는 것에 대해서

f-lab-lyan replied:
https://www.baeldung.com/async-http-client

f-lab-lyan reviewed (approved):
다음 PR에는 테스트를 추가해주시면 좋을 것 같습니다. :-)

canlion committed:
refactor: 회원 관련 DTO-엔티티 변환 메서드를 mapper로 분리

* 여러 종류의 엔티티가 관련된 비즈니스 로직이 추가되며 DTO-엔티티 변환이
  복잡해졌으므로 별도의 mapper를 구현
* 회원 관련 mapper, 컨트롤러에서 공통으로 사용되는 상수를 별도의 상수
  클래스로 관리

canlion committed:
chore: mapper 패키지 구조 단순화

- mapper 수가 적어 패키지를 분리할 필요가 없으므로 단순화

canlion committed:
refactor: 가게 관련 DTO-엔티티 변환 메서드를 mapper로 분리

- 여러 엔티티가 관련된 비즈니스 로직이 추가되어 변환이 복잡해져 별도의
  mapper로 분리함.

canlion committed:
refactor: 카테고리 및 재료 조회 로직 호출을 서비스로 이동

- 카테고리, 재료가 존재하지 않으면 음식 등록 실패함. 따라서 카테고리,
  재료 조회는 비즈니스 로직이므로 서비스 레이어로 이동

canlion committed:
refactor: 음식 관련 DTO-엔티티 변환 메서드를 mapper로 분리

- 여러 엔티티가 관련된 비즈니스 로직이 추가되어 변환이 복잡해져 별도의
  mapper로 분리함.

canlion committed:
refactor: 서비스에 포함된 조회 기능을 룩업 서비스로 이동

- 여러 엔티티를 다루는 서비스가 참조하더라도 순환참조 문제가 발생하지
  않도록 조회 기능을 룩업 서비스에 집중시킴.

canlion committed:
fix: 미완성 상태의 가게 검색 기능을 구현

* 미완성 상태의 가게 검색 기능을 구현
  - 가게, 음식 서비스를 참조하는 퍼사드 서비스 - SearchService 추가
  - 주어진 위경도 범위와 음식 필터를 만족하는 가게와 음식 조회

canlion commented:
## 수정 내역

* 미완성이었던 가게 검색 기능 구현
  - 이전 코드: `StoreController`에서 `StoreService`, `FoodService`를 이용하여 기능 구현을 시도
    - 문제
      - 컨트롤러에 비즈니스 로직이 포함: 단순 서비스 호출을 넘어 반환값에 따라 분기, 결과 후처리까지 로직을 포함합니다.
      - 컨트롤러간의 참조 발생: DTO-엔티티 변환 메서드가 각 엔티티 컨트롤러에 존재하여 가게 검색 결과(`Store`, `Food` 엔티티)를 DTO로 변환하기 위해 컨트롤러간 참조가 필요, 기능이 확장되며 순환참조 가능성이 커집니다.
      - 컨트롤러가 서비스에 종속됨: DTO-엔티티 변환을 컨트롤러에서 수행하므로 컨트롤러는 서비스가 반환한 DTO에 어떤 엔티티들이 포함되어있는지를 알아야하고, 비즈니스 로직이 변경되어 서비스의 반환값이 달라지면 그에 영향을 받습니다.
  - 시도
    - 퍼사드 패턴 적용: `StoreService`, `FoodService`를 이용하여 검색 기능을 담당하는 `SearchService`를 구현
    - 컨트롤러에 포함되어있던 엔티티-DTO 변환을 별도의 mapper로 분리
  - 예상 결과
    - 컨트롤러에서 비즈니스 로직을 제거, 앞서 정한 "컨트롤러는 적절한 서비스를 호출하고 그 결과를 적절한 형태로 변환해 반환한다"는 원칙을 유지할 수 있게 됩니다.
    - 컨트롤러-서비스간 결합을 mapper-서비스간 결합으로 변경. DTO-엔티티 변환을 mapper로 분리하며 서비스의 변경이 컨트롤러에 미치는 영향을 줄입니다. mapper와 서비스간의 결합은 mapper의 역할상 필요한 결합으로 여겨집니다.
* `FoodController`의 엔티티-DTO 변환 로직에 포함된 메뉴카테고리, 재료 조회 메서드 호출을 `FoodService`로 이동
  - 문제
    -  음식 등록 기능 중 메뉴 카테고리, 재료가 등록되지 않은 값인 경우 음식 등록은 실패합니다. 따라서 메뉴 카테고리, 재료 조회와 그에 따른 예외 발생은 비즈니스 로직의 일부라고 생각되며, 그래서 비즈니스 로직 일부가 컨트롤러와 서비스에 나뉘어 위치하고 있습니다.
  - 시도
    - 컨트롤러에서 DTO로 전달받은 메뉴카테고리, 재료를 조회하는 대신 `FoodService.addFood/updateFood` 메서드에 메뉴카테고리, 재료를 파라미터로 추가, 메서드 내부에서 조회하도록 수정
  - 예상 결과
    - 서비스와 컨트롤러에 분할되어 있던 비즈니스 로직을 하나로 정리
  - 기타
    - `FoodService.addFood/updateFood`의 시그니처가 복잡해지면서 컨트롤러-서비스간에도 기본적으로 DTO를 이용하는 것을 고려하고 있습니다.
      - 클라이언트-컨트롤러-엔티티 사이에 서비스를 추가로 끼워넣어 엔티티 노출 실수 방지
      - 서비스 밖에서 발생하는 lazy-loading을 줄임(서비스 내부에서만 DB 엑세스가 가능하도록 한정하면 어플리케이션 성능 추적? 관리?에 유리할 것으로 생각됩니다.)

## 질문

* mapper를 컨트롤러에서 호출하는게 적절한지 모르겠습니다.
  - 적절하다고 생각하는 이유
    - mapper의 호출은 "비즈니스 로직 결과를 클라이언트에 약속한 형태로 변환"하는 것으로 "클라이언트에 약속한 형태"는 컨트롤러의 책임이고 서비스는 알 필요가 없는 지식임.
  - 절적하지 않다고 생각하는 이유
    - (서비스가 엔티티 또는 여러 엔티티를 담은 DTO를 반환하는 상황) mapper가 엔티티를 DTO로 변환하면서 lazy-loading에 따른 DB 엑세스가 컨트롤러에서 발생하고 개발자가 이를 고려해야 함.
    - 서비스에서 미리 필요한 속성들을 fetch하자니 어떤 속성들을 fetch해야하는지 아는 것은 "클라이언트에 약속한 형태"를 알야하므로 책임 범위가 커짐.
  - gpt에 물어봤는데 싸우다가 그만뒀습니다.

f-lab-lyan commented on api/src/main/java/burgermap/mapper/MemberMapper.java:
```diff
@@ -0,0 +1,50 @@
+                .loginId(member.getLoginId())
+                .email(member.getEmail())
+                .nickname(member.getNickname())
+                .profileImageUrl(member.getProfileImage() == null ? null : imageService.getImageUrl(IMAGE_DIRECTORY_PATH, member.getProfileImage().getImageName())
```
DNS를 활용해서 S3와 bucket주소를 감추는 것이 더 일반적인 방법이라고 알고 있어요.

f-lab-lyan reviewed (approved):
수고하셨습니다 👍
